File: stations/system-monitor/tmux_status.py
#!/usr/bin/env python3
"""
tmux_status.py — Unified tmux status bar data provider.

Replaces: sysmon-read.sh, quota-all.sh, quota-ex-segment.sh

Usage:
    python3 tmux_status.py <metric>

Metrics:
    System:  cpu, mem, net, disk, cc, pressure
    Quota:   cc-5h, cc-7d, cc-ex, cx-5h, cx-7d, gm-pro, gm-flash
    Special: ex-segment (styled EX powerline segment)

Read order:
    1. /tmp/agent-metrics-sysmon.json (primary, 5s cache)
    2. HTTP API agent-metrics (fallback)
    3. "?" (graceful fallback)
"""

from __future__ import annotations

import json
import subprocess
import sys
import time
from pathlib import Path
from urllib.error import URLError
from urllib.request import urlopen

PRIMARY = Path("/tmp/agent-metrics-sysmon.json")

# Metric → JSON field mapping
FIELD_MAP = {
    # System metrics (from sysmon-read.sh)
    "cpu": "cpu_display",
    "mem": "mem_display",
    "net": "net_display",
    "disk": "disk_display",
    "cc": "cc_display",
    "pressure": "mem_pressure",
    # LLM quota metrics (from quota-all.sh)
    "cc-5h": "llm_cc_5h",
    "cc-7d": "llm_cc_7d",
    "cc-ex": "llm_cc_ex",
    "cx-5h": "llm_cx_5h",
    "cx-7d": "llm_cx_7d",
    "gm-pro": "llm_gm_pro",
    "gm-flash": "llm_gm_flash",
}

# Quota metrics use different API endpoint and staleness threshold
QUOTA_METRICS = {"cc-5h", "cc-7d", "cc-ex", "cx-5h", "cx-7d", "gm-pro", "gm-flash"}


def _file_age(path: Path) -> float:
    try:
        return time.time() - path.stat().st_mtime
    except OSError:
        return 9999


def _read_from_file(field: str, max_age: float) -> str | None:
    if not PRIMARY.exists():
        return None
    if _file_age(PRIMARY) > max_age:
        return None
    try:
        data = json.loads(PRIMARY.read_text())
        val = data.get(field)
        if val is not None and str(val) not in ("None", ""):
            return str(val)
    except (json.JSONDecodeError, OSError):
        pass
    return None


def _read_from_api(metric: str, field: str) -> str | None:
    if metric in QUOTA_METRICS:
        url = "http://127.0.0.1:8795/quota/formatted"
        key = metric
    else:
        url = "http://127.0.0.1:8795/sysmon/current"
        key = field
    try:
        with urlopen(url, timeout=1) as resp:
            data = json.loads(resp.read())
            val = data.get(key)
            if val is not None and str(val) not in ("None", ""):
                return str(val)
    except (URLError, OSError, json.JSONDecodeError, ValueError):
        pass
    return None


def _tmux_var(name: str) -> str:
    try:
        r = subprocess.run(
            ["tmux", "show", "-gqv", name],
            capture_output=True, text=True, timeout=2,
        )
        return r.stdout.strip()
    except Exception:
        return ""


def get_metric(metric: str) -> str:
    field = FIELD_MAP.get(metric)
    if not field:
        return "?"

    max_age = 120.0 if metric in QUOTA_METRICS else 15.0

    val = _read_from_file(field, max_age)
    if val:
        return val

    val = _read_from_api(metric, field)
    if val:
        return val

    return "?"


def ex_segment() -> str:
    """Conditional EX powerline segment with Catppuccin colors."""
    val = get_metric("cc-ex")
    if not val or val == "?":
        return ""

    flamingo = _tmux_var("@thm_flamingo")
    crust = _tmux_var("@thm_crust")
    fg = _tmux_var("@thm_fg")
    s0 = _tmux_var("@thm_surface_0")
    mantle = _tmux_var("@thm_mantle")

    return (
        f"#[fg={flamingo},bg={mantle}]"
        f"#[fg={crust},bg={flamingo}] EX "
        f"#[fg={fg},bg={s0}] {val} "
        f"#[fg={s0},bg={mantle}]"
    )


def main():
    metric = sys.argv[1] if len(sys.argv) > 1 else "cpu"

    if metric == "ex-segment":
        print(ex_segment(), end="")
    else:
        print(get_metric(metric), end="")


if __name__ == "__main__":
    main()

File: stations/system-monitor-rs/src/tmux_status.rs
//! tmux status line — single line, stdout-only, NO tracing log.
//!
//! Mirrors `stations/system-monitor/tmux_status.py`. Read order:
//!   1. /tmp/agent-metrics-sysmon.json (primary, fast path)
//!   2. HTTP fallback (timeout 1s)
//!   3. "─" placeholder
//!
//! Two top-level kinds dispatched by main.rs:
//!   - "system" → compact `CPU 12% │ MEM 45% │ DISK 67%`
//!   - "llm"    → compact LLM quota glance
//!
//! All 13 sub-metric names from the Python version are also accepted as `kind`
//! so existing tmux.conf `#(... <metric>)` calls keep working.

use anyhow::Result;
use serde_json::Value;
use std::path::Path;
use std::time::Duration;

use crate::config::Settings;

const PRIMARY: &str = "/tmp/agent-metrics-sysmon.json";
const PLACEHOLDER: &str = "─";

/// Map the user-facing metric name to the JSON field written by agent-metrics.
fn field_map(metric: &str) -> Option<&'static str> {
    match metric {
        // System metrics
        "cpu" => Some("cpu_display"),
        "mem" => Some("mem_display"),
        "net" => Some("net_display"),
        "disk" => Some("disk_display"),
        "cc" => Some("cc_display"),
        "pressure" => Some("mem_pressure"),
        // LLM quota metrics
        "cc-5h" => Some("llm_cc_5h"),
        "cc-7d" => Some("llm_cc_7d"),
        "cc-ex" => Some("llm_cc_ex"),
        "cx-5h" => Some("llm_cx_5h"),
        "cx-7d" => Some("llm_cx_7d"),
        "gm-pro" => Some("llm_gm_pro"),
        "gm-flash" => Some("llm_gm_flash"),
        _ => None,
    }
}

fn is_quota(metric: &str) -> bool {
    matches!(
        metric,
        "cc-5h" | "cc-7d" | "cc-ex" | "cx-5h" | "cx-7d" | "gm-pro" | "gm-flash"
    )
}

/// CLI entrypoint. `kind` may be a top-level mode (`system`/`llm`) or a
/// specific sub-metric (e.g. `cpu`, `cc-ex`).
pub async fn print(_cfg: &Settings, kind: &str) -> Result<()> {
    let line = match kind {
        "system" => render_system().await,
        "llm" => render_llm().await,
        // Pass-through single sub-metric.
        other => get_metric(other).await,
    };
    print!("{line}");
    Ok(())
}

async fn render_system() -> String {
    let cpu = get_metric("cpu").await;
    let mem = get_metric("mem").await;
    let disk = get_metric("disk").await;
    format!("CPU {cpu} │ MEM {mem} │ DISK {disk}")
}

async fn render_llm() -> String {
    let cc5 = get_metric("cc-5h").await;
    let cc7 = get_metric("cc-7d").await;
    let ex = get_metric("cc-ex").await;
    if ex == PLACEHOLDER || ex.is_empty() || ex == "off" {
        format!("CC5 {cc5} │ CC7 {cc7}")
    } else {
        format!("CC5 {cc5} │ CC7 {cc7} │ EX {ex}")
    }
}

/// Read a single metric value, walking the file → HTTP → placeholder chain.
async fn get_metric(metric: &str) -> String {
    let Some(field) = field_map(metric) else {
        return PLACEHOLDER.to_string();
    };

    let max_age_secs = if is_quota(metric) { 120.0 } else { 15.0 };
    if let Some(v) = read_from_file(field, max_age_secs) {
        return v;
    }
    if let Some(v) = read_from_api(metric, field).await {
        return v;
    }
    PLACEHOLDER.to_string()
}

fn file_age_secs(path: &Path) -> f64 {
    match std::fs::metadata(path).and_then(|m| m.modified()) {
        Ok(mtime) => mtime
            .elapsed()
            .map(|d| d.as_secs_f64())
            .unwrap_or(f64::INFINITY),
        Err(_) => f64::INFINITY,
    }
}

fn read_from_file(field: &str, max_age: f64) -> Option<String> {
    let path = Path::new(PRIMARY);
    if !path.exists() {
        return None;
    }
    if file_age_secs(path) > max_age {
        return None;
    }
    let raw = std::fs::read_to_string(path).ok()?;
    let data: Value = serde_json::from_str(&raw).ok()?;
    let val = data.get(field)?;
    let s = match val {
        Value::String(s) => s.clone(),
        Value::Null => return None,
        v => v.to_string(),
    };
    if s.is_empty() || s == "None" {
        return None;
    }
    Some(s)
}

async fn read_from_api(metric: &str, field: &str) -> Option<String> {
    // Port resolved from shared/schemas/port_registry.yaml (codegen at build time).
    // Falls back to 10103 if yaml is missing the entry so the binary stays bootable.
    let port = workshop_port_registry::get("agent-metrics")
        .map(|s| s.port)
        .unwrap_or(10103);
    let (url, key) = if is_quota(metric) {
        (format!("http://127.0.0.1:{port}/quota/formatted"), metric)
    } else {
        (format!("http://127.0.0.1:{port}/sysmon/current"), field)
    };
    let client = reqwest::Client::builder()
        .timeout(Duration::from_secs(1))
        .build()
        .ok()?;
    let resp = client.get(&url).send().await.ok()?;
    if !resp.status().is_success() {
        return None;
    }
    let data: Value = resp.json().await.ok()?;
    let val = data.get(key)?;
    let s = match val {
        Value::String(s) => s.clone(),
        Value::Null => return None,
        v => v.to_string(),
    };
    if s.is_empty() || s == "None" {
        return None;
    }
    Some(s)
}
